From 8188ffd534b7d32755a8d6b088a5e63b0bfb593f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?F=C3=A1bio=20Silva?= <fabio@scontain.com>
Date: Wed, 15 Jul 2020 17:46:54 -0300
Subject: [PATCH 1/1] Export patch information

---
 driver/linux/sgx_main.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/driver/linux/sgx_main.c b/driver/linux/sgx_main.c
index 9f3a1ea..d29e352 100644
--- a/driver/linux/sgx_main.c
+++ b/driver/linux/sgx_main.c
@@ -85,6 +85,33 @@ u64 sgx_xfrm_mask = 0x3;
 u32 sgx_misc_reserved;
 u32 sgx_xsave_size_tbl[64];
 
+/*
+ * Patch versions
+ */
+#include <linux/moduleparam.h>
+
+#ifndef PATCH_PAGE0
+#define PATCH_PAGE0 0
+#endif
+
+#ifndef PATCH_METRICS
+#define PATCH_METRICS 0
+#endif
+
+#define IS_DCAP_DRIVER 1
+
+#define COMMIT_SHA "COMMIT_SHA1SUM"
+#define COMMIT_SHA_LEN (40 + 1)
+
+static unsigned int patch_page0 = PATCH_PAGE0;
+static unsigned int patch_metrics = PATCH_METRICS;
+static unsigned int dcap = IS_DCAP_DRIVER;
+static char commit[COMMIT_SHA_LEN] = COMMIT_SHA;
+
+module_param(patch_page0, uint, 0444);
+module_param(patch_metrics, uint, 0444);
+module_param(dcap, uint, 0444);
+module_param_string(commit, commit, COMMIT_SHA_LEN, 0444);
 
 // From intel_sgx.c
 bool sgx_enabled = false;
-- 
2.25.1

