From a0c2cf98e13693e669827abf00d9ae13ede0bf3b Mon Sep 17 00:00:00 2001
From: "pamenas" <mpamenas@gmail.com>
Date: Mon, 15 Feb 2021 19:06:18 +0100
Subject: [PATCH] applied scone dkms patch

---
 dkms.conf | 6 ++++++
 1 file changed, 6 insertions(+)
 create mode 100644 dkms.conf

diff --git a/dkms.conf b/dkms.conf
new file mode 100644
index 0000000..9c1a256
--- /dev/null
+++ b/dkms.conf
@@ -0,0 +1,6 @@
+PACKAGE_NAME="isgx"
+PACKAGE_VERSION="2.11.0"
+BUILT_MODULE_NAME[0]="isgx"
+DEST_MODULE_LOCATION[0]="/kernel/drivers/intel/sgx"
+AUTOINSTALL="yes"
+MAKE[0]="'make'  KDIR=/lib/modules/${kernelver}/build"
-- 
2.17.1

