From 9d6c26e65a428dfc1d1c5723e37ab29658623f3e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?F=C3=A1bio=20Silva?= <fabio@scontain.com>
Date: Wed, 2 Sep 2020 14:27:49 -0300
Subject: [PATCH] Add DKMS support

---
 dkms.conf | 6 ++++++
 1 file changed, 6 insertions(+)
 create mode 100644 dkms.conf

diff --git a/dkms.conf b/dkms.conf
new file mode 100644
index 0000000..9c03f4a
--- /dev/null
+++ b/dkms.conf
@@ -0,0 +1,6 @@
+PACKAGE_NAME="isgx"
+PACKAGE_VERSION="2.6.0"
+BUILT_MODULE_NAME[0]="isgx"
+DEST_MODULE_LOCATION[0]="/kernel/drivers/intel/sgx"
+AUTOINSTALL="yes"
+MAKE[0]="'make'  KDIR=/lib/modules/${kernelver}/build"
-- 
2.25.1

